File: pr-3553/sphinx/snippets/cpp/using/examples/hadamard_gate.cpp
/*******************************************************************************
 * Copyright (c) 2025 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// [Begin Docs]
#include <cudaq.h>
#include <iostream>

__qpu__ void hadamard_gate() {
  // Allocate a qubit.
  cudaq::qubit q;
  // Apply a hadamard gate to the qubit.
  h(q);
  // Measure the qubit.
  mz(q);
}

int main() {
  // Call the kernel to run the hadamard gate.
  auto result = cudaq::sample(hadamard_gate);

  // Default number of shots
  int shots = 1000;
  int count0 = result.count("0");
  int count1 = result.count("1");

  // result.dump();
  std::cout << "Measured |0> with probability "
            << static_cast<double>(count0) / shots << std::endl;
  std::cout << "Measured |1> with probability "
            << static_cast<double>(count1) / shots << std::endl;
  return 0;
}
// [End Docs]

File: pr-3929/sphinx/snippets/python/using/examples/hadamard_gate.py
# ============================================================================ #
# Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

#[Begin Docs]
import cudaq


@cudaq.kernel
def kernel():
    # A single qubit initialized to the ground/ zero state.
    qubit = cudaq.qubit()

    # Apply Hadamard gate to single qubit to put it in equal superposition.
    h(qubit)

    # Measurement operator.
    mz(qubit)


result = cudaq.sample(kernel)
print("Measured |0> with probability " +
      str(result["0"] / sum(result.values())))
print("Measured |1> with probability " +
      str(result["1"] / sum(result.values())))
#[End Docs]
